### ISSUE TITLE
Need push-based commit/branch metadata to select correct test case versions

### ISSUE THREAD

**Issue by ameyer1024** [utgheith/grader#15]
Currently, the grader uses `git rev-list` to select the correct commit to use for the test case, but this is fragile when the history has merges.

The command is approximately this:
```sh
git rev-list -n 1 --first-parent --until {test_cutoff} master
```
In this, we use `--first-parent` as an approximate heuristic to figure out what chain of commits was the "main branch" (since the usually commits are merged from other branches into the main branch, leaving the main branch as the first parent) but that isn't always the case.

In the most recent failure case (test `001` on `cs439t_f24_p4`), this was the commit graph:
```
* 0f0fa83 (HEAD -> master, origin/master, origin/HEAD)
*   ba9a4f6 Merge remote-tracking branch 'origin'
|\  
| * 40e9b3b
| * 8d0421d
| * 6b25321
| * 190f147
| * 8229bdc
| * 056ac01
| * 55f035e
* | 90c2379
|/  
* c23f1b8
* f54ad03
* 3421a35
* 84d7b20
```

Up until today, `master` pointed to the chain of commits preceding `40e9b3b`, which had the correct updated test case; however, the merge commit `ba9a4f6` had the existing master branch as the second parent rather than the first, so after that point, `git log --first-parent master` only saw `ba9a4f6`, `90c2379`, `c23f1b8`... and not the original master branch.

Basically, the `--first-parent` heuristic doesn't work, and I think the only actual way to solve this is to track which commit each branch referenced at each push, and then use the commit on the master branch from the last push before the deadline.  This would fully solve the issue of being able to accidentally change which commit the test case grader selects by changing the way in which git log/ref-log traverses the history graph (such as by adding later merge commits).
